### ISSUE TITLE
Fix clobbering other people's saves in the editor

### ISSUE THREAD

**Issue by jaltekruse** [Doenet/DoenetTools#1890]
Proposal for first level quick fix - Optimistic concurrency control

-   when you save, don't just give the new content you want to store at
    a given name/ID

    -   also send the version number (document ID + server timestamp of
        save) of the document that you pulled down when you started
        editing

        -   Or because we do frequent auto-saves it would actually be
            whatever last version you know was saved to the server
            successfully

    -   If you have a "last known saved version" in your request that is
        earlier than the last known saved version the server knows
        about, reject the save and show a message to the user that
        someone has modified the document since your last save

    -   Super basic version

        -   That is the end of the feature, this new server validation
            and user warning

        -   a user that sees this error has the option to open the
            document in a new tab and do manual conflict resolution

            -   We could give them a convenience link to open the tab

    -   Nicer version

        -   Pull down the newer version from the server and show it to
            them in a diff viewer, and make them resolve the conflicts
            manually before saving their changes

        -   If they really don't want to resolve the conflicts right now
            and just want to continue on their current editing path,
            they have the option of creating a temporary document,
            copying their version out of the one side of the diff viewer
            and pasting it into the document of a different name to
            enable continued editing

        -   If they want to re-incorporate their work back into the
            shared document that can do that later manually

-   For concurrent editors on a DoenetML, I think there is a bit of a
    natural forcing function of a structured text format that will limit
    the usefulness of actual real-time collaborative editing. As you
    edit you are constantly making some edits that temporarily break the
    structure of the XML until you are done with an edit (and maybe even
    a few clicks of the render button that report where you had
    errors/typos), mixing that together with another person editing
    somewhere else concurrently seems pretty annoying to deal with.

    -   I think any automatic merging of changes will require a more
        complex system of version management

        -   In our meeting we discussed only doing merging when at least
            you have a valid parsable XML structure, but there isn't an
            obvious solution to handle the fact you might leave your
            work in a broken state for hours/days until you can get back
            to it to finish your latest update, in which case you could
            significantly diverge from other editors while we wait for
            the condition that would allow us to make an effort to
            automatically merge things.

        -   Also some edits of an XML document to restructure bigger
            elements of the activity could produce diffs that would very
            likely require a human to resolve the differences

        -   It seems likely in my mind (Jason) that a real time editor
            would be best done when we have a structured graphical
            editor, so the high level structure can never be invalid,
            and the conflicts that would arise would be logical
            conflicts like referring to a name in a copy tag that had
            been renamed by another editor

            -   But mixing structural errors that come up while you are
                halfway through typing a tag or moving around and
                reorganizing an XML document into constant real-time
                transparent merging seems a bit like trying to code
                together in a shared IDE, you would be really annoyed if
                your debugging and test runs of the code could be messed
                up if someone else was halfway through writing a
                function and constantly making the code not
                compile/parse

    -   A short term solution that at least prevents clobbering and
        warns a user that they need to resolve the conflicts themselves
        should handle our current most pressing current problems (if I
        understand them correctly)

**Comment by jaltekruse** [comment]
@kevincharles and @Lyanthropos

**Comment by dqnykamp** [comment]
I recommend checking out how overleaf does real time editing with latex.  They have to deal with the same issues that we do, where intermediate edits can leave the document with invalid latex that will not compile.  My guess is that they simply don't worry about it and just merge edit immediately.  At least, you can see the cursor of the other person and that's what it looks like they are doing.  I've never paid much attention, though.

I feel like if such a system works for them, it should work for us.

**Comment by jaltekruse** [comment]
@dqnykamp it looks like they have a permissive latex parser, I can have a document with errors in it and it will still render what it does understand.

I'm not proposing that optimistic concurrency control is a long term solution, but I'm guessing you were responding to the medium to long term discussion about possibly needing a graphical editor to do realtime. If we have a permissive parser I could see it working reasonably like overleaf does.

**Comment by dqnykamp** [comment]
Right, I was responding to the medium/long term discussion.  I hope we could get it working reasonably with a text editor like overleaf does.  We certainly do not have a permissive parser at present; that hadn't been my philosophy.  But, it's worth revisiting if it will help with the real time.  And maybe it is a better user experience.  (Overleaf certainly has taken that approach.)

The permissive parser question is a piece of the broader question of how we should deal with errors in the DoenetML.

**Comment by jaltekruse** [comment]
Implemented the basic version with optimistic concurrency control in https://github.com/Doenet/DoenetTools/pull/2088
